ARROW-82: Initial IPC support for ListArray

emkornfield opened the pull request:
This is a work in progress because I can't get clang-tidy to shut-up on parameterized test files (see last commit which I need to revert before merge).   I'd like to make sure this is a clean build and make sure people are ok with these change.  This PR also has a lot of collateral damage for small/large things I cleaned up my way to make this work.  I tried to split the commits up logically but if people would prefer separate pull requests I can try to do that as well.

Open questions:
1.  For supporting strings, binary, etc.  I was thinking of changing thei4 type definitions to inherit from ListType, and to hard-code the child type.  This would allow for simpler IPC code (all of the instantiation of types would happen in construct.h/.cc?) vs handling each of there types separately for IPC.
2.  There are some TODOs I left sprinkled in the code and would like peoples thoughts on if they are urgent/worthwhile for following up on.

Open issues:
1.  Supporting the rest of the List backed logical types
2.  More unit tests for added functionality.

As part of this commit I also refactored the Builder interfaces a little bit for the following reasons:
1.  It seems that if ArrayBuilder owns the bitmap it should be responsible for having methods to manipulate it.
2.  This allows ListBuilder to use the parent class + a BufferBuilder instead of inheriting Int32Builder, which means it doesn't need to do strange length/capacity hacks.

Other misc things here:
1.  Native popcount in test-util.h
2.  Ability to build a new list on top of an existing by incrementally add offsets/sizes
3.  Added missing types primitive types in construct.h for primitive.

wesm commented:
Awesome, I will review in detail when I can.

Also pinging @jacques-n on code reviews -- can we figure out a code review solution that is less horrible than GitHub (I hate to be so negative but: really)? I'd love to use gerrithub.io but may require blessing from ASF Infra. I could ask Cloudera folks to set up a public Gerrit for us, too, but I'd rather figure out something that runs under the ASF umbrella.

jacques-n commented:
Our options are really: 

1. Reviewboard (official Apache solution)
2. Github (somewhat Apache supported)
3. External Gerrit run by someone (that pushes to mailing list) similar to what Kudu does.

See here for more sadness: https://issues.apache.org/jira/browse/INFRA-2205

I'm inclined towards (3) and fall back to (1). (2) is actually the worse in my experience.

wesm commented:
Are you all (in particular @emkornfield) cool if I ask for a gerrit.cloudera.org to be set up? I asked ASF Infra for GerritHub but I'm not that optimistic. 

If there is another way we can set up a Gerrit for Arrow that would be great too. The ultimate would be getting the ASF to manage it (unfortunately, people have been "losing karma" for using external code review tools).

emkornfield commented:
I don't understand all the implications for moving to Gerrit.   A few questions (this might be a better topic for the ML?):
1.  Is the desire for Gerrit simply as a better UI, or does it improve workflows in some way (looking at the ticket Jacques referenced it seems like the answer is "Yes")?
2.  Does gerrit.cloudera.org start handling the CI?   Does it make it easier to do things, like tracking performance benchmarks (I remember the cribbed run-tests.sh, mentioning failed builds get uploaded someplace else)?  Any notion of how much work it is to modify our CI to work with Gerrit?  How much on going support is required from cloudera for modifying pre-commit hooks?
3.  Does adding another tool into the mix discourage or add friction to contributions?

On a UI level, I have a preference for ReviewBoard over Gerrit but that might just be a matter of familiarity. I've never used Gerrit in a code review workflow only perused a few open source projects using it, so I suppose it might grow on me.

emkornfield commented on cpp/src/arrow/builder.cc:
```diff
@@ -36,6 +55,7 @@ Status ArrayBuilder::Init(int32_t capacity) {
 }
 
 Status ArrayBuilder::Resize(int32_t new_bits) {
```
By the way I was thinking it would be nice to make Resize and Init virtual methods, I think it would reduce code repetition and potential bugs for classes the don't have any reason to override Reserve.  Thoughts?

wesm replied:
Sounds good to me, go ahead

wesm commented:
To your questions, briefly:

1) it's a significantly improved workflow. Gerrit makes large code reviews transactional -- it's much more organized and things don't fall through the cracks like they do in GitHub. 

2) it'd be easier in the short term to keep running builds with Travis CI -- so there would be some finagling to push updated patches to a PR branch as well as Gerrit (for reviews)

3) I don't think we would need to require Gerrit for anything right now, so patches would still start with a pull request and we could have reviews for larger patches (> 50 lines or whatever) go to Gerrit if it will be deemed helpful.

wesm commented:
Let me open a ML discussion about this as I'm sure others will have opinions.

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -25,6 +25,25 @@
+}
+
+Status ArrayBuilder::AppendToBitmap(const uint8_t* valid_bytes, int32_t length) {
+  Reserve(length);
```
`Reserve` may return error status

emkornfield replied:
Oops.  I wonder how hard it would be write a clang-tidy rule to always ensure methods with a Status field are checked.

wesm replied:
Kudu wrote an LLVM static analyzer for this https://github.com/apache/incubator-kudu/tree/master/build-support/tools/kudu-lint

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -56,10 +76,46 @@ Status ArrayBuilder::Advance(int32_t elements) {
 
 Status ArrayBuilder::Reserve(int32_t elements) {
   if (length_ + elements > capacity_) {
+    // TODO(emkornfield) power of 2 growth is potentially suboptimal
```
Aside: should we do 1.5x growth everywhere (this is the folly approach IIRC -- is there more research on this subject?)

emkornfield replied:
Yes this is what folly uses.  Seems like 1.5 edges out 2 on the very light survey done here: https://en.wikipedia.org/wiki/Dynamic_array#Growth_factor 

Ideally, we would benchmark once we have some real data in place.  In the absence of that 1.5 seems like a good default.

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -56,10 +76,46 @@ Status ArrayBuilder::Advance(int32_t elements) {
+  }
+  for (int32_t i = 0; i < length; ++i) {
+    // TODO(emkornfield) Optimize for large values of length?
+    AppendToBitmap(valid_bytes[i] == 0);
```
Should this be UnsafeAppendToBitmap? This presumes that the user has called `Reserve(...)` beforehand

emkornfield replied:
correct.  I will change.

wesm commented on cpp/src/arrow/array.h:
```diff
@@ -39,7 +40,7 @@ class Array {
       const std::shared_ptr<Buffer>& null_bitmap = nullptr);
 
-  virtual ~Array() {}
+  virtual ~Array() = default;
```
Aside: I tried to find definitive documentation on when to use `= default` vs `= 0` in C++11, if you have any pointers

emkornfield replied:
My understanding is default == {}, and destructor = 0, should be used rarely.  From: http://en.cppreference.com/w/cpp/language/destructor "A destructor may be declared pure virtual, for example in a base class which needs to be made abstract, but has no other suitable functions that could be declared pure virtual. Such destructor must have a definition, since all base class destructors are always called when the derived class is destroyed"  Although the second sentence seems to contradict the first.

emkornfield replied:
I changed to default purely for style (I likely should have done this as a separate PR and discussed), for some reason it reads more elegantly to me.

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -25,6 +25,25 @@
 
 namespace arrow {
 
+Status ArrayBuilder::AppendToBitmap(bool is_null) {
```
As a matter of consistency, should this be `is_valid`?

emkornfield replied:
Makes sense, I did it this way because it was closer to how the code previously worked.  But I will change it.

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -56,10 +76,46 @@ Status ArrayBuilder::Advance(int32_t elements) {
+  return Status::OK();
+}
+
+void ArrayBuilder::UnsafeAppendToBitmap(bool is_null) {
```
same question as above

emkornfield replied:
yes, I think so.  I will make the change.

wesm commented on cpp/src/arrow/ipc/adapter.cc:
```diff
@@ -63,32 +65,56 @@ static bool IsPrimitive(const DataType* type) {
+    // the types behave the same way?
+    // case Type::BINARY:
+    // case Type::CHAR:
+    case Type::LIST:
```
See also @jacques-n comment on the mailing list. Having logical types for String (UTF8) and Binary makes sense, but internally these are represented in memory using the same memory layout as List<UInt8>. So how we implement them in C++ should be whatever is the most convenient and high performance thing, probably

wesm commented on cpp/src/arrow/ipc/adapter.cc:
```diff
@@ -63,32 +65,56 @@ static bool IsPrimitive(const DataType* type) {
-      // Push a dummy zero-length buffer, not to be copied
-      buffers->push_back(std::make_shared<Buffer>(nullptr, 0));
-    }
+  const auto arr_type = arr->type().get();
```
Declare this type

wesm commented on cpp/src/arrow/ipc/adapter.cc:
```diff
@@ -63,32 +65,56 @@ static bool IsPrimitive(const DataType* type) {
-    }
+  const auto arr_type = arr->type().get();
+  if (IsPrimitive(arr_type)) {
+    const PrimitiveArray* prim_arr = static_cast<const PrimitiveArray*>(arr);
```
auto is appropriate here, as below

wesm commented on cpp/src/arrow/ipc/adapter.cc:
```diff
@@ -63,32 +65,56 @@ static bool IsPrimitive(const DataType* type) {
-    // TODO(wesm)
-    return Status::NotImplemented("List type");
+  } else if (IsListType(arr_type)) {
+    const ListArray* list_arr = static_cast<const ListArray*>(arr);
```
auto is OK here

wesm commented on cpp/src/arrow/ipc/adapter.cc:
```diff
@@ -63,32 +65,56 @@ static bool IsPrimitive(const DataType* type) {
+  } else if (IsListType(arr_type)) {
+    const ListArray* list_arr = static_cast<const ListArray*>(arr);
+    buffers->push_back(list_arr->offset_buffer());
+    // TODO(emkornfield) limit recursion depth
```
Good point. Impala limited to 32K depth IIRC, but at minimum we should track the depth and have a static limit.

wesm commented on cpp/src/arrow/ipc/adapter.cc:
```diff
@@ -226,15 +252,17 @@ class RowBatchReader::Impl {
+    // extract null_bitmap which is common to all arrays
+    std::shared_ptr<Buffer> null_bitmap;
+    if (field_meta.null_count == 0) {
+      null_bitmap = nullptr;
```
is this redundant?

emkornfield replied:
I don't think so.  The way the spec is currently worded it says it is optional on whether null_count == 0 implies a null pointer, and that the presence of it should be communicated in the metadata.  I need to check Message.fbs, but I'm not sure it currently allows for communicating whether the bitmap is included.  Here we make the decision to never include the null bitmap because the data-structure is redundant.  I will add a comment explaining this.

wesm replied:
Ah sorry, I meant that the nullary ctor for `shared_ptr` sets nullptr (versus having to do so explicitly)

emkornfield replied:
yes good point.

wesm commented on cpp/src/arrow/ipc/adapter.cc:
```diff
@@ -243,6 +271,26 @@ class RowBatchReader::Impl {
       return MakePrimitiveArray(
           type, field_meta.length, data, field_meta.null_count, null_bitmap, out);
     }
+    if (IsListType(type.get())) {
```
else if?

emkornfield replied:
I will check what clang-tidy says.  In other cases it liked not having else if when the original "if" block always returned.   If we prefer to allow either I can look into turning off that particular check.

tnachen replied:
Either else if, or I also think having more consistency in spacing will be better, so extra space after closing braces.

wesm replied:
The prior block is executed if `IsPrimitive(...)` is true -- in theory they should be mutually exclusive

wesm commented on cpp/src/arrow/ipc/adapter.cc:
```diff
@@ -243,6 +271,26 @@ class RowBatchReader::Impl {
+      if (field_meta.length > 0) {
+        RETURN_NOT_OK(GetBuffer(buffer_index_++, &offsets));
+      } else {
+        offsets.reset(new Buffer(nullptr, 0));
```
This is interesting -- I suppose with length-0 `ListArray` instances, there is no need to have any offsets allocated at all (versus a length-1 array with a single element 0).

emkornfield replied:
yeah, I wasn't sure, its not mentioned in the spec.  I'll send another e-mail out to the ML about this.

emkornfield replied:
After further consideration I think we should always serialize the offsets. I don't think it is worth consuming code have to special case for 0 length lists.

wesm commented on cpp/src/arrow/ipc/adapter.cc:
```diff
@@ -243,6 +271,26 @@ class RowBatchReader::Impl {
+        offsets.reset(new Buffer(nullptr, 0));
+      }
+      const int num_children = type->num_children();
+      if (num_children != 1) {
```
You could skip the local variable creation (to avoid an extra `type->num_children()` below), but I'm ambivalent.

wesm commented on cpp/src/arrow/ipc/ipc-adapter-test.cc:
```diff
@@ -40,25 +39,62 @@
 namespace ipc {
 
-class TestWriteRowBatch : public ::testing::Test, public MemoryMapFixture {
+// TODO(emkornfield) convert to google style kInt32, etc?
```
Seems OK to me, this is a reminder for me to start doing this too

emkornfield replied:
I'm going to open JIRA to try to do a universal cleanup of these and enums.

wesm commented on cpp/src/arrow/ipc/ipc-adapter-test.cc:
```diff
@@ -40,25 +39,62 @@
+    std::shared_ptr<RowBatchReader> reader;
+    RETURN_NOT_OK(RowBatchReader::Open(mmap_.get(), header_location, &reader));
+
+    // TODO(emkornfield): why does this require a smart pointer for schema?
```
the returned `RowBatch` retains a `shared_ptr<Schema>`

wesm commented on cpp/src/arrow/ipc/ipc-adapter-test.cc:
```diff
@@ -67,42 +103,40 @@ TEST_F(TestWriteRowBatch, IntegerRoundTrip) {
+// TODO(emkornfield) More tests
+// Test primitive and lists with zero elements
+// Tests lists and primitives with no nulls
+// String  type
```
I'm OK with these leaking over to new JIRAs or some other place to keep track

emkornfield replied:
I think I will add tests to cover the first two and file jira to better support string (and refactor the logical string types to make it convenient for reusing the list serialization logic).  Sound reasonable?

wesm replied:
Sounds good

wesm commented on cpp/src/arrow/type.h:
```diff
@@ -116,7 +116,7 @@ struct DataType {
   bool Equals(const DataType* other) {
     // Call with a pointer so more friendly to subclasses
-    return this == other || (this->type == other->type);
+    return this == other || ((other != nullptr) && (this->type == other->type));
```
You could do `other && (this == other || ..)` but I don't have a preference

emkornfield replied:
Interesting note (a very knowledgeable C++ pointed this out to me):
```
DataType* t = nullptr
assert_true(t->Equals(nullptr))
```
will work work (and not segfault).  I believe your suggestion will return false.  Of course once we get to that point we might want undefined behavior anyways.

emkornfield replied:
and by my logic I should also be checking this != nullptr

wesm replied:
Fair enough. I am generally comfortable turning a blind eye on the more pedantic points in the interest of productivity

emkornfield replied:
agree.  And so does the compiler actually, when I tried adding in the this != nullptr, it complains that this should never be null in well defined c++ code.  I switched it to your suggestion.

wesm commented on cpp/src/arrow/types/construct.cc:
```diff
@@ -90,11 +89,43 @@ Status MakePrimitiveArray(const std::shared_ptr<DataType>& type, int32_t length,
+    case Type::CHAR:
+    case Type::DECIMAL_TEXT:
+    case Type::STRING:
+    case Type::VARCHAR:
```
Aside: We should figure out whether these variable-length "primitive" types will be considered list-like within this library or more fundamentally first-class primitive types

emkornfield replied:
Yes, it might pay to pull all this code into one method that takes the superset of all parameters needed to make an Array to avoid having to make the distinctions for issue like this.  Part of me cringes a little bit at this though.

wesm replied:
I guess my question was more of the split between physical memory representation and logical types. While String / UTF8 has the same memory representation as List<UInt8>, internally to arrow-C++ we may consider it to be a primitive type (and same in the metadata)

wesm commented:
Reviewed the code, seems in good shape and thanks for much cleanup. I will address your other questions / comments in the morning (European time)

tnachen commented on cpp/src/arrow/array.h:
```diff
@@ -58,6 +59,9 @@ class Array {
   virtual bool Equals(const std::shared_ptr<Array>& arr) const = 0;
+  // Determines if the array is internally consistent.  Defaults to always
+  // returning Status::OK.  This can be an expensive check.
+  virtual Status Validate() const;
```
Pardon my dumb question, my understanding of Status just seeing its usage in the code I thought it's signaling either an operation succeeded or failed, but validate seems to imply it's either in a valid state or invalid state (like a bool). Why not just use bool? Or does Status can also encode error information?

emkornfield replied:
Status can encode error information.

wesm replied:
There was some work last year in Impala in which they analyzed the generated x86 instructions to see how to minimize the CPU cycles associated with verifying Status::OK: https://github.com/apache/incubator-impala/commit/1afe72830a696e700ac5948ac6e2e0994c50f991#diff-ca5b27977d768460882519dbb9162a4e

emkornfield commented:
Thanks.  I should be able to push a commit addressing all the comments by the end of the weekend.

emkornfield commented:
OK, I think I addressed all of the suggestions here.  I filed JIRAs for remaining issues.  I accidentally ran clang-format-3.8 which re-arranged some of the headers to make them alphabetical.   This lead to some header reordering which looks correct.  some other stuff that disagrees with clang-format-3.7 so I had to push one additional commit to make the check-format pass.  Sorry for the noisy code review.

wesm commented on cpp/src/arrow/ipc/ipc-adapter-test.cc:
```diff
@@ -123,20 +117,140 @@ Status MakeListRowBatch(std::shared_ptr<RowBatch>* out) {
+  ASSERT_OK(RowBatchReader::Open(mmap_.get(), header_location, &reader));
+  std::shared_ptr<RowBatch> batch_result;
+  ASSERT_RAISES(Invalid, reader->GetRowBatch(schema, &batch_result));
+}
```
Thank you for these thorough tests =)

wesm commented:
Great, thank you for all of this additional work. At some point I figure we'll move to LLVM 3.8 -- for now I guess we can use include-what-you-use or other tools for deeper header cleanup.

+1, merging now
